### PR TITLE
haskell: make generic builder follow compiler’s shared config

### PR DESCRIPTION
**Pull request by matthewbauer**
enableShared in generic-builder.nix should default to what the GHC
compiler was compiled with. Add a passthru to all of the GHC compilers
to hold the value of enableShared. If enableShared is not set in the
GHC we just use false as the default value for enableSharedLibraries.

Note: I may have missed some compilers. Only GHC & GHCJS are covered
by this commit but this shouldn’t break evaluation of anything else.

### CODE REVIEW

**Review by Ericson2314 — APPROVED** [review]

**Comment by Ericson2314** [comment]
Added 8.2.1-binary

**Comment by peti** [comment]
> enableShared in generic-builder.nix should default to what the GHC compiler was compiled with.

Why should it?

**Comment by matthewbauer** [comment]
> Why should it?

My understanding is that if you're compiler wasn't compiled with "enableShared", setting "enableSharedLibraries" to true, will break things.

**Comment by Ericson2314** [comment]
```
enableSharedLibraries ? (ghc.isGhcjs or false) || stdenv.lib.versionOlder "7.7" ghc.version)
```
is always true today. I gather this is because old GHC didn't support dynamic linking, so that reads "true where possible". If anyone has a custom compiler that doesn't support dynamic linking, they can do `passthru = { enableShared = false; };` just like with the prebuilt ones. If we want to add additional conditions we can do
```
enableSharedLibraries ? ghc.enableShared or false && additionalCondition
```
The point being deduplicating the logic like this doesn't tie our hands.

**Comment by peti** [comment]
> My understanding is that if your compiler wasn't compiled with "enableShared", setting "enableSharedLibraries" to true, will break things.

I see an argument to be made that you'd *want* things to break in that case. The `enableShared` default in the generic builder can be viewed as expressing a policy that compilers ought to fulfill and if someone uses the generic builder with a compiler that does not fulfill that policy then you ought to pass a proper `enableShared` setting along your custom compiler. That approach makes these choices more *explicit*, whereas the new scheme that @Ericson2314 chose to merge without waiting for my review makes this choice implicit.

**Comment by matthewbauer** [comment]
Feel free to revert if you disagree. This was motivated by the need to have "enableSharedLibraries" set to false in prebuilt Android & iOS (which is the default for their config) and they are broken otherwise. I felt like this approach would avoid the most duplication but there are certainly other ways to accomplish this. If you do want to revert I hope it is okay to add an extra condition to enableSharedLibraries in generic-builder.nix.

**Comment by Ericson2314** [comment]
Yeah we can revert this in the interim, but let me continue to argue for this approach.

@peti so first of all, are you OK with us adding the cross exceptions to the generic builder default? If we do, then the default becomes a verbose condition that doesn't to me express a policy other than "build shared libraries where possible". (Any other interpretation would strike me as an "overfit" one.) If we don't, then users must manually instruct generic-builder even when they don't actually care and just want working builds.

**Comment by peti** [comment]
I just would have liked to have this conversation *before* the merge, not after. Anyway, I don't feel strongly enough about this issue to bother reverting those changes now. Both scheme have their advantages, and the state we have now feels just fine.

**Pull request merged**
